src/util/mp_arith.h

```diff
@@ -62,82 +64,93 @@ unsigned integer2unsigned(const mp_integer &);
 
 const mp_integer mp_zero=string2integer("0");
 
+class exprt;
+class constant_exprt;
+class typet;
+
+// this one will go away
+// returns 'true' on error
+/// \deprecated: use the constant_exprt version instead
+bool to_integer(const exprt &expr, mp_integer &int_value);
+
+// returns 'true' on error
+/// \deprecated: use numeric_cast<mp_integer> instead
+bool to_integer(const constant_exprt &expr, mp_integer &int_value);
+
+// returns 'true' on error
+bool to_unsigned_integer(const constant_exprt &expr, unsigned &uint_value);
+
 /// Numerical cast provides a unified way of converting from one numerical type
 /// to another.
 /// Generic case doesn't exist, this has to be specialized for different types.
-template <typename Target, typename Source, typename = void>
+template <typename Target, typename = void>
 struct numeric_castt final
 {
 };
 
+template <>
+struct numeric_castt<mp_integer> final
+{
+  optionalt<mp_integer> operator()(const exprt &expr) const
+  {
+    mp_integer out;
+    if(to_integer(expr, out))
+      return {};
+    return out;
+  }
+};
+
 /// Convert mp_integer to any signed type
 /// \tparam T: type to convert to
 /// \param mpi: mp_integer to convert
 /// \return optional integer of type T if conversion is possible,
 ///         empty optional otherwise.
 template <typename T>
 struct numeric_castt<T,
-                     mp_integer,
-                     typename std::enable_if<std::is_integral<T>::value &&
-                                             std::is_signed<T>::value>::type>
+                     typename std::enable_if<std::is_integral<T>::value>::type>
 {
-  static optionalt<T> numeric_cast(const mp_integer &mpi)
+  template <typename U = T,
+            typename std::enable_if<std::is_signed<U>::value, int>::type = 0>
+  static auto get_val(const mp_integer &mpi) -> decltype(mpi.to_long())
+  {
+    return mpi.to_long();
+  }
+
+  template <typename U = T,
+            typename std::enable_if<!std::is_signed<U>::value, int>::type = 0>
+  static auto get_val(const mp_integer &mpi) -> decltype(mpi.to_ulong())
+  {
+    return mpi.to_ulong();
+  }
+
+  optionalt<T> operator()(const mp_integer &mpi) const
   {
+    constexpr const auto precondition =
```

Comment:
Surely this will be a static error on VS2013 which doesn't have constexprs in the standard library

Reply:
Hmm, potentially yeah. You could put an ifdef around the constexpr I suppose, or we might have a macro that conditionally expands to constexpr in util somewhere.